_posts/2022-09-07-storage-write-removal-before-conditional-termination.md

```diff
@@ -0,0 +1,133 @@
+
+If this termination of the transaction happens conditionally within a function, i.e. the function can either successfully terminate the transaction or return (as in exit to continue control flow after the call at the call site), the optimizer may incorrectly remove state changes before a call to the function.
+
+This happens in cases, in which the state change before the function could be considered redundant, if the fact that the function can terminate the transaction was ignored.
+
+When using the legacy code generator, the Yul optimizer step responsible for the bug will be run on each inline assembly block in isolation and hence the above conditions have to occur inside one inline assembly block. While still possible, this makes the occurrence of the bug with legacy code generation less likely (see the section of Technical Details below for examples of affected assembly snippets).
+
+However, the new via-IR code generation translates the entire contract to Yul and runs the optimizer on all of it. In this case, the bug can be triggered more easily. The following is a minimal example:
+
+```
+contract C {
+	uint public x;
+	function f() public {
+		x = 1; // This write used to be removed by the yul optimizer.
+		g();
+		x = 2;
+	}
+	function g() internal {
+		// The important aspect of this function is that it can both return to the caller and terminate the transaction.
+		// The bug will show its effects in the cases in which it terminates the transaction, since the state change above will be missing.
+		if (msg.data.length > 4) return;
+		assembly { return(0,0) }
+	}
+}
+```
+
+Compiling the above via-IR with enabled optimizer will result ``f()`` without additional calldata to incorrectly return without modifying ``x``.
+
+For a concrete contract, the bug may still be prevented, in case the optimizer inlines the function before running the problematic store removal step, e.g. the following is almost
+identical, but unaffected, since ``g()`` will be inlined (in the snippet above the ``return;`` prevents easy inlining):
+
+```
+contract C {
+	uint public x;
+	function f() public {
+		x = 1;
+		g();
+		x = 2;
+	}
+	function g() internal {
+		if (msg.data.length <= 4)
+			assembly { return(0,0) }
+	}
+}
+```
+
+
+## Technical Details
+
+On the Yul level the issue manifests if the following sequence occurs:
+
+- A storage write occurs (e.g. ``sstore(0, 1)``).
+- A user-defined Yul function is called and this function both has a control flow path terminates the transaction and a control flow path out of the function.
+- Another storage write occurs that makes the first storage write above redundant (e.g. ``sstore(0, 2)``).
```

Comment:
I feel like this should actually be in "Which Contracts are Affected?" and a lot of what is there now should be here in "Technical Details". If I wanted to check if my contract is affected without going into details, I'd probably want to see such a concise list of conditions needed to trigger the bug.

Reply:
The problem is that this sequence is only accurate for this happening on the yul level during optimization. Which patterns in solidity can lead to it is more complicated... that's why I'd consider this rather a technical detail than the usual conditions for affected contracts...